src/ServiceControl/DbMigrations/1.27/SplitFailedMessageDocumentsMigration.cs

```diff
@@ -1,156 +1,179 @@
                 .ToArray();
 
-            var retries = processingAttempts.Where(x => x.IsRetry).ToArray();
+            //Split the original FailedMessage into separate documents based on new unique message id
+            var failedMessages = processingAttempts
+                .GroupBy(p => p.UniqueMessageId)
+                .Select(g => new FailedMessage
+                {
+                    Id = FailedMessage.MakeDocumentId(g.Key),
+                    UniqueMessageId = g.Key,
+                    ProcessingAttempts = g.OrderBy(a => a.Index).Select(a => a.Attempt).ToList(),
+                    Status = FailedMessageStatus.Unresolved
+                }).ToList();
+
+            //Do nothing if we don't split the document
+            if (failedMessages.Count == 1) return stats;
 
-            var nonRetries = processingAttempts.Except(retries).ToArray();
+            stats.FoundProblem++;
 
-            if (nonRetries.Length <= 1)
+            if (failedMessages.All(f => f.UniqueMessageId != originalFailedMessage.UniqueMessageId))
             {
-                return stats;
+                session.Delete(originalFailedMessage);
+                stats.Deleted++;
             }
+            else
+            {
+                var failedMessageCopy = failedMessages.Single(f => f.UniqueMessageId == originalFailedMessage.UniqueMessageId);
+                failedMessages.Remove(failedMessageCopy);
+                failedMessages.Add(originalFailedMessage);
 
-            writer.WriteLine($"{failedMessage.UniqueMessageId} {failedMessage.Status}");
-            stats.FoundProblem = 1;
+                originalFailedMessage.ProcessingAttempts = failedMessageCopy.ProcessingAttempts;
+                originalFailedMessage.Status = failedMessageCopy.Status;
+            }
 
-            foreach (var attempt in processingAttempts)
+            failedMessages.ForEach(failedMessage =>
             {
-                var retryMarker = attempt.IsRetry ? "R" : "A";
+                var lastAttempt = failedMessage.ProcessingAttempts.Last();
 
-                writer.WriteLine($"\t{retryMarker}: {attempt.MessageId} {attempt.MessageType} {attempt.FailedQ}");
-            }
+                object messageType;
+
+                if (lastAttempt.MessageMetadata.TryGetValue("MessageType", out messageType))
+                {
+                    if (!failedMessage.FailureGroups.Any())
+                    {
+                        failedMessage.FailureGroups = failedMessageFactory.GetGroups((string)messageType, lastAttempt.FailureDetails);
+                    }
+                    failedMessage.FailureGroups.Add(CreateSplitFailureGroup(lastAttempt, messageType, originalStatus));
+                }
 
-            writer.WriteLine();
+                if (failedMessage.UniqueMessageId == originalFailedMessage.UniqueMessageId) return;
+
+                session.Store(failedMessage);
+                stats.Created++;
```

Comment:
This counts the original document as well if it is still around. I'm not opposed to that, just noting it.